httpx/middleware/timeout.go

```diff
@@ -70,21 +69,30 @@ func (h *timeoutHandler) ServeHTTPContext(ctx context.Context, w http.ResponseWr
 		err = h.handler.ServeHTTPContext(ctx, tw, r)
 		close(done)
 	}()
+
 	select {
 	case p := <-panicChan:
 		panic(p)
 	case <-done:
 		tw.mu.Lock()
 		defer tw.mu.Unlock()
-		dst := w.Header()
-		for k, vv := range tw.h {
-			dst[k] = vv
-		}
-		if !tw.wroteHeader {
-			tw.code = http.StatusOK
+
+		// If timeout writer was written to by request handler, we write the buffered
+		// response to the response writer.
+		//
+		// It is possible that the handler merely returned an error in which case
+		// a middleware may write a response, so we must not always write one here.
+		if tw.modified || len(tw.Header()) > 0 {
```

Comment:
it maybe better to move this logic to `timeoutWriter` as something like `IsModified()` so that it can control what defines being modified.